Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+        g_variant_builder_init(&actionsBuilder, G_VARIANT_TYPE("as"));
+
+        if (m_capabilities.contains(Capabilities::Actions)) {
+            g_variant_builder_add(&actionsBuilder, "s", "default");
+            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
+        }
+
+        const char* applicationID = nullptr;
+        if (auto* app = g_application_get_default())
+            applicationID = g_application_get_application_id(app);
+
+        GVariantBuilder hintsBuilder;
+        g_variant_builder_init(&hintsBuilder, G_VARIANT_TYPE("a{sv}"));
+        if (applicationID)
+            g_variant_builder_add(&hintsBuilder, "{sv}", "desktop-entry", g_variant_new_string(applicationID));
+
+        auto* value = static_cast<GValue*>(fastZeroedMalloc(sizeof(GValue)));
+        g_value_init(value, G_TYPE_UINT64);
+        g_value_set_uint64(value, notification.notificationID());
+
+        CString body;
+        if (m_capabilities.contains(Capabilities::Body))
+            body = notification.body().utf8();
+
+        const char* applicationIcon = nullptr;
+#if PLATFORM(GTK)
+        if (auto* iconTheme = gtk_icon_theme_get_for_display(gdk_display_get_default()))
+            applicationIcon = applicationID && gtk_icon_theme_has_icon(iconTheme, applicationID) ? applicationID : nullptr;
+#endif
+
+        g_dbus_proxy_call(m_proxy.get(), "Notify", g_variant_new(
+            "(susssasa{sv}i)",
+            g_get_application_name(), addResult.iterator->value.id, applicationIcon ? applicationIcon : "",
+            notification.title().utf8().data(), body.data(),
+            &actionsBuilder, &hintsBuilder, -1
+            ), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, [](GObject* source, GAsyncResult* result, gpointer userData) {
+                GUniqueOutPtr<GError> error;
+                GRefPtr<GVariant> notificationID = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(source), result, &error.outPtr()));
+                if (!notificationID)
+                    g_warning("Failed to show notification: %s", error->message);
+
+                guint32 id;
+                g_variant_get(notificationID.get(), "(u)", &id);
+
+                auto* value = static_cast<GValue*>(userData);
+                NotificationService::singleton().setNotificationID(g_value_get_uint64(value), id);
+                g_value_unset(value);
+                fastFree(value);
+            }, value);
+    }
+    return true;
+}
+
+void NotificationService::cancelNotification(uint64_t webNotificationID)
+{
+    if (!m_proxy)
+        return;
+
+    auto it = m_notifications.find(webNotificationID);
+    if (it == m_notifications.end())
+        return;
+
+    if (shouldUsePortal()) {
+        if (it->value.portalID.isEmpty())
+            return;
+
+        g_dbus_proxy_call(m_proxy.get(), "RemoveNotification", g_variant_new("(s)", it->value.portalID.utf8().data()), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```

Comment:
I'd like to see callbacks everywhere you omit them currently. Even doing nothing except print error and return is valuable.